62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+	}
+};
+
+Hatch::QuadraticBezier Hatch::Segment::splitCurveRange(QuadraticBezier curve, double left, double right) const
+{
+    return curve.splitCurveTakeLower(right).splitCurveTakeUpper(left / right);
+}
+
+Hatch::QuadraticBezier Hatch::Segment::getSplitCurve() const
+{
+    return splitCurveRange(*originalBezier, t_start, t_end);
+}
+
+bool Hatch::Segment::isStraightLineConstantMajor() const
+{
+	auto major = (uint)SelectedMajorAxis;
+	return originalBezier->p[0][major] == originalBezier->p[1][major] &&
+		originalBezier->p[0][major] == originalBezier->p[2][major];
+}
+
+std::array<double, 2> Hatch::Segment::intersect(const Segment& other) const
+{
+	auto major = (int)SelectedMajorAxis;
+	auto minor = 1-major;
+	// Use line intersections if one or both of the beziers are linear (a = 0)
+	if (originalBezier->isLineSegment())
+	{
+		if (other.originalBezier->isLineSegment())
+		{
+			// Line/line intersection
+			// TODO: make this less shit
+			auto x1 = originalBezier->p[0][minor], y1 = originalBezier->p[0][major],
+				x2 = originalBezier->p[2][minor], y2 = originalBezier->p[2][major],
+				x3 = other.originalBezier->p[0][minor], y3 = other.originalBezier->p[0][major],
+				x4 = other.originalBezier->p[2][minor], y4 = other.originalBezier->p[2][major];
+			auto nx = (x1 * y2 - y1 * x2) * (x3 - x4) - (x1 - x2) * (x3 * y4 - y3 * x4), 
+				ny = (x1 * y2 - y1 * x2) * (y3 - y4) - (y1 - y2) * (x3 * y4 - y3 * x4),
+				d = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4);
+			auto intersectionX = nx / d, intersectionY = ny / d;
+
+			// Return if point is on the lines
+			if (d != 0.0 && 
+				std::min(x1, x2) <= intersectionX && y1 <= intersectionY && std::max(x1, x2) >= intersectionX && y2 >= intersectionY &&
+				std::min(x3, x4) <= intersectionX && y3 <= intersectionY && std::max(x3, x4) >= intersectionX && y4 >= intersectionY)
+				return { other.originalBezier->intersectOrtho(ny / d, major), core::nan<double>() };
+
+			return { core::nan<double>(), core::nan<double>() };
+		}
+
+		// Line/curve intersection
+		// TODO: make this less shit as well
+		auto tx = originalBezier->p[0][minor], ty = originalBezier->p[0][major],
+			a = -atan2(originalBezier->p[2][minor] - tx, originalBezier->p[2][major] - ty);
+		QuadraticBezier rotatedOther = {
+			float64_t2((other.originalBezier->p[0].x - tx)* cos(a) - (other.originalBezier->p[0].y - ty) * sin(a), (other.originalBezier->p[0].x - tx) * sin(a) + (other.originalBezier->p[0].y - ty) * cos(a)),
+			float64_t2((other.originalBezier->p[1].x - tx) * cos(a) - (other.originalBezier->p[1].y - ty) * sin(a), (other.originalBezier->p[1].x - tx) * sin(a) + (other.originalBezier->p[1].y - ty) * cos(a)),
+			float64_t2((other.originalBezier->p[2].x - tx) * cos(a) - (other.originalBezier->p[2].y - ty) * sin(a), (other.originalBezier->p[2].x - tx) * sin(a) + (other.originalBezier->p[2].y - ty) * cos(a))
+		};
+		return { rotatedOther.intersectOrtho(0, major), core::nan<double>() };
```

Comment:
beware if you implement my pseudo code in [previous comment](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/52/files#r1368675921)  then you should do `intersectOrtho(0.0, 1)` cause you're always rotation to align with x axis